packages/react-components/react-utilities/src/compose/resolveShorthand.ts

```diff
@@ -1,17 +1,16 @@
 import { isValidElement } from 'react';
 import type { ReplaceNullWithUndefined, SlotShorthandValue, UnknownSlotProps } from './types';
 
-export type ResolveShorthandOptions<Props, Required extends boolean = false> = {
-  required?: Required;
-  defaultProps?: Props;
-};
+export type ResolveShorthandOptions<Props> =
+  | { required: true; defaultProps?: Props }
+  | { required?: false; defaultProps?: Props };
```

Comment:
Technically this is a breaking change since this type is exported and you're removing a type parameter. I think you can get by without using this type at all (see comment below). Perhaps you could then leave this as-is (maybe add `@deprecated`).